src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx

```diff
@@ -0,0 +1,226 @@
+import { FormattedMessage, FormattedMessage as Msg, useIntl } from 'react-intl';
+
+import getSurveysWithElements from 'features/smartSearch/fetching/getSurveysWithElements';
+import { COLUMN_TYPE, SelectedViewColumn } from '../types';
+import {
+  ELEMENT_TYPE,
+  RESPONSE_TYPE,
+  ZetkinSurveyElement,
+} from 'utils/types/zetkin';
+
+interface SurveyResponseConfigProps {
+  onOutputConfigured: (columns: SelectedViewColumn[]) => void;
+}
+
+export enum SURVEY_QUESTION_OPTIONS {
+  ALL_OPTIONS = 'AllSingle',
+  ALL_OPTIONS_SEPARATED = 'AllOptionsSeparated',
+  ONE_OPTION = 'oneOption',
+}
+
+const SurveyResponseConfig = ({
+  onOutputConfigured,
+}: SurveyResponseConfigProps) => {
+  const { orgId } = useRouter().query;
+  const surveysQuery = useQuery(
+    ['surveysWithElements', orgId],
+    getSurveysWithElements(orgId as string)
+  );
+  const intl = useIntl();
+  const columnsReset: SelectedViewColumn[] = [];
+  const [surveyId, setSurveyId] = useState<number | null>();
+  const [selectedQuestion, setSelectedQuestion] =
+    useState<ZetkinSurveyElement | null>(null);
+
+  const onSurveyChange: ChangeEventHandler<{ value: unknown }> = (ev) => {
+    setSurveyId(ev.target.value as number);
+    setSelectedQuestion(null);
+    onOutputConfigured(columnsReset);
+  };
+
+  const makeColumnsText = (selectedElement: ZetkinSurveyElement) => {
+    if (selectedElement.question.response_type === RESPONSE_TYPE.TEXT) {
+      return [
+        {
+          config: {
+            question_id: selectedElement.id,
+          },
+          title: selectedElement.question.question,
+          type: COLUMN_TYPE.SURVEY_RESPONSE,
+        },
+      ];
+    }
+  };
+
+  const makeColumnsOption = (
+    selectedOption: string | SURVEY_QUESTION_OPTIONS
+  ) => {
+    if (selectedQuestion !== undefined && selectedQuestion !== null) {
+      if (selectedOption === SURVEY_QUESTION_OPTIONS.ALL_OPTIONS) {
+        return [
+          {
+            config: {
+              question_id: selectedQuestion.id,
+            },
+
+            title: selectedQuestion.question.question,
+            type: COLUMN_TYPE.SURVEY_OPTIONS,
+          },
+        ];
+      } else if (
+        selectedOption === SURVEY_QUESTION_OPTIONS.ALL_OPTIONS_SEPARATED
+      ) {
+        return selectedQuestion.question.options?.map((option) => {
+          return {
+            config: {
+              option_id: option.id,
+              survey_id: selectedQuestion.id,
+            },
+            title: option.text,
+            type: COLUMN_TYPE.SURVEY_OPTION,
+          };
+        });
+      } else {
+        const optionTitle = selectedQuestion.question.options?.find(
+          (option) => option.id === (selectedOption as unknown as number)
+        );
+        if (optionTitle === undefined) {
+          return [];
+        }
+        return [
+          {
+            config: {
+              option_id: selectedOption as unknown as number,
```

Comment:
i get really curious about this type casting and assume there is a story here?

Reply:
Yes, this is problematic. The reason for this cast is that `option_id` expects a `number`, but according to the function argument list, `selectedOption` is either a `string` or a `SURVEY_QUESTION_OPTION`. My guess is that @rebecarubio has tried to cast it as `number` directly, but there is no overlap between the `string | SURVEY_QUESTION_OPTION` union and `number`, so Typescript has recommended that it be cast "through" `unknown`.

The core issue here is that either the types are wrong (and `selectedOption` is really a `number`) or in this branch (`else`) the type of `selectedOption` is `string`, in which case the code is lying to typescript, which fails to discover that we're using a `string` where a `number` is expected, which would result in the wrong type of data being sent to the server. In that case, using `parseInt(selectedOption as string)` would have been a more correct thing to do.

When I look in the browser network panel to see what's actually getting sent to the server, I see a `number` here, so my analysis is that he types are wrong. The correct type of the `selectedOption` should probably be `number | SURVEY_QUESTION_OPTION`.